// @spectrum-web-components/overlay/src/overlay-trigger-directive.dev.js
"use strict";
import {
  nothing,
  render
} from "@spectrum-web-components/base";
import { directive } from "lit/async-directive.js";
import { Overlay, strategies } from "./Overlay.dev.js";
import "../sp-overlay.dev.js";
import {
  removeSlottableRequest
} from "./slottable-request-event.dev.js";
import { SlottableRequestDirective } from "./slottable-request-directive.dev.js";
export class OverlayTriggerDirective extends SlottableRequestDirective {
  constructor() {
    super(...arguments);
    this.overlay = new Overlay();
    this.defaultOptions = {
      triggerInteraction: "click",
      overlayOptions: {
        placement: "top-start",
        type: "auto",
        offset: 0
      }
    };
    this.options = {
      ...this.defaultOptions.overlayOptions
    };
  }
  /* c8 ignore next 9 */
  render(_template, _options) {
    return nothing;
  }
  update(part, [template, options]) {
    var _a, _b, _c;
    this.options = {
      ...this.defaultOptions.overlayOptions,
      ...options == null ? void 0 : options.overlayOptions
    };
    this.insertionOptions = options == null ? void 0 : options.insertionOptions;
    this.template = template;
    let newTarget = false;
    const triggerInteraction = (options == null ? void 0 : options.triggerInteraction) || this.defaultOptions.triggerInteraction;
    const newStrategy = ((_a = this.strategy) == null ? void 0 : _a.type) !== triggerInteraction;
    if (this.target !== part.element) {
      this.target = part.element;
      newTarget = true;
    }
    if (newTarget || newStrategy) {
      (_b = this.strategy) == null ? void 0 : _b.abort();
      this.strategy = new strategies[triggerInteraction](this.overlay, this.target, true);
    }
    this.listenerHost = this.overlay;
    this.init();
    this.overlay.open = (_c = options == null ? void 0 : options.open) != null ? _c : false;
  }
  handleSlottableRequest(event) {
    var _a, _b;
    if (event.target !== event.currentTarget)
      return;
    const willRemoveSlottable = event.data === removeSlottableRequest;
    render(willRemoveSlottable ? void 0 : this.template(), this.overlay);
    if (willRemoveSlottable) {
      this.overlay.remove();
    } else {
      Overlay.applyOptions(this.overlay, {
        ...this.options,
        trigger: this.target
      });
      const insertionEl = typeof ((_a = this.insertionOptions) == null ? void 0 : _a.el) === "function" ? this.insertionOptions.el() : ((_b = this.insertionOptions) == null ? void 0 : _b.el) || this.target;
      const { where = "afterend" } = this.insertionOptions || {};
      insertionEl.insertAdjacentElement(where, this.overlay);
    }
  }
}
export const trigger = directive(OverlayTriggerDirective);
